README.md
# Block Merge Based on Time

Some organizations have the policy to prohibit their members from merging pull requests because it would be difficult to
handle troubles after working hours or on holidays. This action helps them to ensure such a policy in enforcement.

Block Merge Based on Time
[creates a commit status](https://docs.github.com/en/rest/reference/repos#create-a-commit-status) for each pull request
head commit. If the time is during **blocking**, it creates a commit status with **pending**. On the other hand, it
creates a commit status with **success** while it's not the time to block. You can circumvent this block with a specific
label named `no-block` by default. This might be useful when you have to merge pull requests because of an emergency or
irregular maintenance.

## Usage

Create a new workflow like this:

```yaml
name: Block Merge Based on Time
on:
  pull_request:
    types:
      - opened
      - reopened
      - synchronize
      - labeled
      - unlabeled
  schedule:
    - cron: "*/30 * * * *"

jobs:
  block:
    runs-on: ubuntu-latest
    steps:
      - uses: yykamei/block-merge-based-on-time@main
        with:
          timezone: Pacific/Honolulu
          after: "17:30, 16:30 on Monday"
          before: 09:00
          base-branches: "(default)"
          prohibited-days-dates: "Sunday, 2021-10-01, 2021-12-29/2022-01-04, H:United States, BH:United States"
```

### Action inputs

These are all available inputs.

| Name                                       | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | Required | Default                                                                            |
| ------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------- | ---------------------------------------------------------------------------------- |
| `timezone`                                 | Time zone to use. Default is UTC                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | `true`   | -                                                                                  |
| `after`                                    | The time to start blocking merge. You can set exception time for specific days with comma-separated values. For example, the value could be "17:30, 16:30 on Monday, 15:00 on Friday." You can omit the base hour with specific days, so "16:30 on Monday" is also valid. Note you should omit the base hour from both of `after` and `before` because it assumes the base hour as `"00:00"` it they are omitted. If only one parameter include the base hour, you will unexpected behavior.                                                                              | `false`  | -                                                                                  |
| `before`                                   | The time to stop blocking merge. You can set exception time for specific days with comma-separated values. For example, the value could be "09:00, 08:00 on Monday" You can omit the base hour with specific days, so "08:00 on Monday" is also valid. Note you should omit the base hour from both of `after` and `before` because it assumes the base hour as `"00:00"` it they are omitted. If only one parameter include the base hour, you will unexpected behavior.                                                                                                 | `false`  | -                                                                                  |
| `base-branches`                            | The comma-separated base branches of pull requests. This tool will block only pull requests the base branches of which are listed in this parameter. You can set regular expressions surrounding with `/` like `/staging\/.*/` and just set string literals like `develop`. Also, the value `(default)` is acceptable, which means pull requests that will be merged into the default branch will be blocked. The default value is `/^.*$/` for backward compatibility.                                                                                                   | `false`  | `"/^.*$/"`                                                                         |
| `prohibited-days-dates`                    | The comma-separated days or dates to stop blocking merge for all day. You can also specify regional holidays with special syntax, such as "H:St. Barthélemy" and "BH:St. Barthélemy", which stand for "holidays of St. Barthélemy" and "before holidays of St. Barthélemy." The word after "H:" or "BH:" is a region name that is listed in [`src/holidays.json`](https://raw.githubusercontent.com/yykamei/block-merge-based-on-time/main/src/holidays.json) as a JSON key. For example, the value could be "Sunday, 2021-08-01, 2021-08-06/2021-08-10, H:Côte d’Ivoire" | `false`  | `""`                                                                               |
| `no-block-label`                           | The label to indicate the pull request should not be blocked                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | `false`  | `no-block`                                                                         |
| `commit-status-context`                    | The commit status context                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | `false`  | `block-merge-based-on-time`                                                        |
| `commit-status-description-with-success`   | The commit status description shown with success                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | `false`  | `The PR could be merged`                                                           |
| `commit-status-description-while-blocking` | The commit status description shown while blocking                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | `false`  | `The PR can't be merged based on time, which is due to your organization's policy` |
| `commit-status-url`                        | The commit status URL to describe why this action is conducted                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            | `false`  | `""`                                                                               |
| `token`                                    | The GitHub token used to create an authenticated client                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   | `false`  | `GITHUB_TOKEN`                                                                     |

## Regional holidays

This tool supports blocking merges based on regional holidays. For example, February 11th is "National Foundation Day"
in Japan, and you can block merges on such a holiday via `prohibited-days-dates` with `H:Japan`. In addition, you can
also block on the day before the regional holiday with `BH:Japan`.

Block Merge Based on Time supports **232** regions, taking advantage of Google Calendar API, and the data is updated
periodically.
See [`src/holidays.json`](https://raw.githubusercontent.com/yykamei/block-merge-based-on-time/main/src/holidays.json) to
check all available regions.

## Contributing

Please take a look at
the [CONTRIBUTING.md](https://github.com/yykamei/block-merge-based-on-time/blob/main/CONTRIBUTING.md). It's always a
pleasure to receive any contributions 😄


CHANGELOG.md
# Changelog

## [v2.1.21](https://github.com/yykamei/block-merge-based-on-time/tree/v2.1.21) (2023-03-25)

[Full Changelog](https://github.com/yykamei/block-merge-based-on-time/compare/v2.1.20...v2.1.21)

**Implemented enhancements:**

- Update Holidays [\#1142](https://github.com/yykamei/block-merge-based-on-time/pull/1142) ([github-actions[bot]](https://github.com/apps/github-actions))

**Merged pull requests:**

- Bump @types/node from 18.15.3 to 18.15.9 [\#1147](https://github.com/yykamei/block-merge-based-on-time/pull/1147) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump @typescript-eslint/eslint-plugin from 5.55.0 to 5.56.0 [\#1146](https://github.com/yykamei/block-merge-based-on-time/pull/1146) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump prettier from 2.8.4 to 2.8.7 [\#1145](https://github.com/yykamei/block-merge-based-on-time/pull/1145) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump @typescript-eslint/parser from 5.55.0 to 5.56.0 [\#1144](https://github.com/yykamei/block-merge-based-on-time/pull/1144) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump eslint-config-prettier from 8.7.0 to 8.8.0 [\#1143](https://github.com/yykamei/block-merge-based-on-time/pull/1143) ([dependabot[bot]](https://github.com/apps/dependabot))

## [v2.1.20](https://github.com/yykamei/block-merge-based-on-time/tree/v2.1.20) (2023-03-24)

[Full Changelog](https://github.com/yykamei/block-merge-based-on-time/compare/v2.1.19...v2.1.20)

**Implemented enhancements:**

- Update Holidays [\#1141](https://github.com/yykamei/block-merge-based-on-time/pull/1141) ([github-actions[bot]](https://github.com/apps/github-actions))

## [v2.1.19](https://github.com/yykamei/block-merge-based-on-time/tree/v2.1.19) (2023-03-23)

[Full Changelog](https://github.com/yykamei/block-merge-based-on-time/compare/v2.1.18...v2.1.19)

**Implemented enhancements:**

- Update Holidays [\#1140](https://github.com/yykamei/block-merge-based-on-time/pull/1140) ([github-actions[bot]](https://github.com/apps/github-actions))

## [v2.1.18](https://github.com/yykamei/block-merge-based-on-time/tree/v2.1.18) (2023-03-18)

[Full Changelog](https://github.com/yykamei/block-merge-based-on-time/compare/v2.1.17...v2.1.18)

**Implemented enhancements:**

- Update Holidays [\#1134](https://github.com/yykamei/block-merge-based-on-time/pull/1134) ([github-actions[bot]](https://github.com/apps/github-actions))

**Merged pull requests:**

- Bump @types/node from 18.15.0 to 18.15.3 [\#1137](https://github.com/yykamei/block-merge-based-on-time/pull/1137) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump @typescript-eslint/eslint-plugin from 5.54.1 to 5.55.0 [\#1136](https://github.com/yykamei/block-merge-based-on-time/pull/1136) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump @typescript-eslint/parser from 5.54.1 to 5.55.0 [\#1135](https://github.com/yykamei/block-merge-based-on-time/pull/1135) ([dependabot[bot]](https://github.com/apps/dependabot))

## [v2.1.17](https://github.com/yykamei/block-merge-based-on-time/tree/v2.1.17) (2023-03-15)

[Full Changelog](https://github.com/yykamei/block-merge-based-on-time/compare/v2.1.16...v2.1.17)

**Implemented enhancements:**

- Update Holidays [\#1131](https://github.com/yykamei/block-merge-based-on-time/pull/1131) ([github-actions[bot]](https://github.com/apps/github-actions))

## [v2.1.16](https://github.com/yykamei/block-merge-based-on-time/tree/v2.1.16) (2023-03-11)

[Full Changelog](https://github.com/yykamei/block-merge-based-on-time/compare/v2.1.15...v2.1.16)

**Implemented enhancements:**

- Update Holidays [\#1125](https://github.com/yykamei/block-merge-based-on-time/pull/1125) ([github-actions[bot]](https://github.com/apps/github-actions))

**Merged pull requests:**

- Bump @typescript-eslint/eslint-plugin from 5.54.0 to 5.54.1 [\#1130](https://github.com/yykamei/block-merge-based-on-time/pull/1130) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump @types/node from 18.14.6 to 18.15.0 [\#1129](https://github.com/yykamei/block-merge-based-on-time/pull/1129) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump @typescript-eslint/parser from 5.54.0 to 5.54.1 [\#1128](https://github.com/yykamei/block-merge-based-on-time/pull/1128) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump eslint-config-prettier from 8.6.0 to 8.7.0 [\#1127](https://github.com/yykamei/block-merge-based-on-time/pull/1127) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump eslint from 8.35.0 to 8.36.0 [\#1126](https://github.com/yykamei/block-merge-based-on-time/pull/1126) ([dependabot[bot]](https://github.com/apps/dependabot))

## [v2.1.15](https://github.com/yykamei/block-merge-based-on-time/tree/v2.1.15) (2023-03-10)

[Full Changelog](https://github.com/yykamei/block-merge-based-on-time/compare/v2.1.14...v2.1.15)

**Implemented enhancements:**

- Update Holidays [\#1124](https://github.com/yykamei/block-merge-based-on-time/pull/1124) ([github-actions[bot]](https://github.com/apps/github-actions))

**Closed issues:**

- Cron job is not running on opened PRs [\#1123](https://github.com/yykamei/block-merge-based-on-time/issues/1123)

**Merged pull requests:**

- Bump luxon from 3.2.1 to 3.3.0 [\#1122](https://github.com/yykamei/block-merge-based-on-time/pull/1122) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump @types/node from 18.14.1 to 18.14.6 [\#1121](https://github.com/yykamei/block-merge-based-on-time/pull/1121) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump @typescript-eslint/parser from 5.53.0 to 5.54.0 [\#1120](https://github.com/yykamei/block-merge-based-on-time/pull/1120) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump eslint from 8.34.0 to 8.35.0 [\#1119](https://github.com/yykamei/block-merge-based-on-time/pull/1119) ([dependabot[bot]](https://github.com/apps/dependabot))

src/inputs.ts
import { getInput } from "@actions/core"
import type { Zone } from "luxon"
import { DateTime, Interval } from "luxon"
import holidays from "./holidays.json"
import type { Dates, Days, DaysDates, HolidayEntry, Hours } from "./types"

export class Inputs {
  public readonly token: string
  public readonly after: Hours
  public readonly before: Hours
  public readonly timezone: Zone
  public readonly prohibitedDays: Days
  public readonly prohibitedDates: Dates
  public readonly noBlockLabel: string
  public readonly commitStatusContext: string
  public readonly commitStatusDescriptionWithSuccess: string
  public readonly commitStatusDescriptionWhileBlocking: string
  public readonly commitStatusURL: string | null
  private readonly rawBaseBranches: string[]

  constructor() {
    this.token = getInput("token", { required: true })
    this.timezone = timeZone()
    this.after = hours("after", this.timezone)
    this.before = hours("before", this.timezone)
    const [days, dates] = prohibitedDaysDates(this.timezone)
    this.prohibitedDays = days
    this.prohibitedDates = dates
    this.noBlockLabel = stringOr(getInput("no-block-label"), "no-block")
    this.commitStatusContext = stringOr(getInput("commit-status-context"), "block-merge-based-on-time")
    this.commitStatusDescriptionWithSuccess = stringOr(
      getInput("commit-status-description-with-success"),
      "The PR could be merged"
    )
    this.commitStatusDescriptionWhileBlocking = stringOr(
      getInput("commit-status-description-while-blocking"),
      "The PR can't be merged based on time, which is due to your organization's policy"
    )
    // NOTE: If the string is empty, we're not sure where we should refer to. So, `||` is appropriate here instead of `??`.
    this.commitStatusURL = getInput("commit-status-url") || null
    this.rawBaseBranches = getInput("base-branches")
      .split(/,\s*/)
      .filter((v) => v !== "")
  }

  public baseBranches(defaultBranch: string): RegExp[] {
    return this.rawBaseBranches.map((v) => {
      if (v === "(default)") {
        return new RegExp(`^${escapeRegExpCharacters(defaultBranch)}$`)
      } else if (v.startsWith("/") && v.endsWith("/")) {
        return new RegExp(v.replace(/^\/(.*)\/$/, "$1"))
      }
      return new RegExp(`^${escapeRegExpCharacters(v)}$`)
    })
  }
}

function timeZone(): Zone {
  const d = DateTime.now().setZone(getInput("timezone"))
  if (d.invalidExplanation != null) {
    throw new Error(d.invalidExplanation)
  }
  return d.zone
}

function dateTime(format: string, s: string, zone: Zone): DateTime {
  const d = DateTime.fromFormat(s, format, { zone, setZone: true })
  if (d.invalidExplanation != null) {
    throw new Error(d.invalidExplanation)
  }
  return d
}

function interval(s: string, zone: Zone): Interval {
  let ret: Interval
  if (s.split("/", 2).length === 2) {
    ret = Interval.fromISO(s, { zone })
    if (ret.end != null) {
      ret = ret.set({ end: ret.end.endOf("day") })
    }
  } else {
    const start = dateTime("yyyy-MM-dd", s, zone)
    ret = Interval.fromDateTimes(start, start.endOf("day"))
  }
  if (ret.invalidExplanation != null) {
    throw new Error(ret.invalidExplanation)
  }
  return ret
}

function hours(key: "after" | "before", zone: Zone): Hours {
  const input = getInput(key)
  const baseRegExp = /^\d\d:\d\d$/
  const daysRegExp = /^(?<hour>\d\d:\d\d) on (?<day>Sunday|Monday|Tuesday|Wednesday|Thursday|Friday|Saturday)$/
  const result: Hours = input.split(/,\s*/).reduce(
    (result, str) => {
      if (baseRegExp.test(str)) {
        return { ...result, base: dateTime("hh:mm", str, zone) }
      } else {
        const match = str.match(daysRegExp)
        if (match != null && match.groups != null && match.groups["day"] != null && match.groups["hour"] != null) {
          const day = match.groups["day"]
          const hour = match.groups["hour"]
          return { ...result, [day]: dateTime("hh:mm", hour, zone) }
        } else {
          throw new Error(`Invalid "${key}" was given. The example format is "16:30 on Monday"`)
        }
      }
    },
    { base: dateTime("hh:mm", "00:00", zone) }
  )
  return result
}

function holidayEntries(region: string): HolidayEntry[] {
  const validRegion = (r: string): r is keyof typeof holidays => r in holidays
  if (validRegion(region)) {
    return holidays[region]
  }
  throw new Error(`Unsupported region is given: "${region}"`)
}

function prohibitedDaysDates(zone: Zone): DaysDates {
  const days: Days = []
  const dates: Dates = []
  getInput("prohibited-days-dates")
    .split(/,\s*/)
    .forEach((s) => {
      switch (s) {
        case "Sunday":
        case "Monday":
        case "Tuesday":
        case "Wednesday":
        case "Thursday":
        case "Friday":
        case "Saturday": {
          days.push(s)
          break
        }
        case "":
          break // If the input is empty string, split array will have one empty string in it.
        default: {
          if (s.startsWith("H:")) {
            const [_prefix, region] = s.split("H:", 2)
            if (region != null) {
              holidayEntries(region).forEach((entry) => {
                dates.push(interval(entry.date, zone))
              })
            }
          } else if (s.startsWith("BH:")) {
            const [_prefix, region] = s.split("BH:", 2)
            if (region != null) {
              holidayEntries(region).forEach((entry) => {
                let d = DateTime.fromISO(entry.date)
                d = d.set({ day: d.day - 1 })
                dates.push(interval(d.toISODate(), zone))
              })
            }
          } else {
            dates.push(interval(s, zone))
          }
          break
        }
      }
    })
  return [days, dates]
}

/**
 *
 * return the passed string as is but an alternative when the passed string is empty.
 *
 * @param {string} str
 * @param {string} alt
 */
function stringOr(str: string, alt: string): string {
  return str || alt
}

// NOTE: I followed this guide.
//       https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions#escaping
function escapeRegExpCharacters(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")
}
